feat(engine): Cold Start - Primary Parameters

Taz5150 opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
Fixes #5001
Fixes #4621 
Fixes #2351

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
This PR features a realistic cold engine start-up process for the primary parameters (N1, N2, EGT and FF). Engine spools as a function of pressure and temperature until Idle condition is reached.

No additional visuals have been implemented/ changed, except for the small delay in the start-valve opening (SD Engine page).

Sound XML file updated and adapted to new LVars (courtesy of boris#8008).

New Engine LVars included in the FDR code (courtesy of AGuther [Z+2]#0001).

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
Before engine start, all parameters are stable at 0 except EGT which is set at ambient temperature.

![EWD-1](https://user-images.githubusercontent.com/70781153/121480736-ac3aa280-c9cb-11eb-8442-6f8e1a735918.PNG)

Expected behavior of N2, EGT and FF as a function of time during start-up:

![image](https://user-images.githubusercontent.com/70781153/121481589-924d8f80-c9cc-11eb-8f2e-f801f57af93f.png)


## References
<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->
- Several YT videos of real Leap start-ups
- Discord user FlyingCarpet7#7502

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username for Engine Model (if different from GitHub): TazX [Z+2]#0405
Discord username for Sound Model (if different from GitHub): boris#8008

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
For testing, check that the start-up procedure is as expected in the figures above for different set of environmental conditions:

- High and low altitude airports
- High and low ambient temperatures

Other than the process itself:
- check that EGT initial temperature starts at ambient temperature
- check that the transition from start-up to Idle is smooth (since two separate functions are being used)

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

MisterChocker commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2003,19 +2005,15 @@ var A320_Neo_UpperECAM;
         getN1GaugeValue() {
             const engineId = (this.index + 1);
-            const value = SimVar.GetSimVarValue("ENG N1 RPM:" + engineId, "percent");
+            const value = SimVar.GetSimVarValue("L:A32NX_ENGINE_N1:" + engineId, "percent");
+            if (value < 0.0) {
+                value = 0;
+            }
```
This doesn't work, you'll need to use a let instead.

Taz5150 replied:
Done! Thanks.

aguther commented on flybywire-aircraft-a320-neo/SimObjects/AirPlanes/FlyByWire_A320_NEO/engines.cfg:
```diff
@@ -107,4 +109,4 @@ thrust_scalar = 1
 DisableMixtureControls = 1 ; Disable mixture controls
 DisableParkingBrakeControls = 0 ; Disable parking brake controls
-DisablePropellerControls = 0 ; Disable propeller controls
+DisablePropellerControls = 0 ; Disable propeller controls
```
avoid unnecessary changes

Taz5150 replied:
Not sure what happened there, honestly.

aguther commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2198,15 +2196,8 @@ var A320_Neo_UpperECAM;
-            }
-            const name = "ENG N2 RPM:" + _engine;
-            let percent = SimVar.GetSimVarValue(name, "percent") - n2Imbalance;
+            const name = "L:A32NX_ENGINE_N2:" + _engine;
```
creation of name can be done inline like for N1

Taz5150 replied:
Change committed.

aguther reviewed (commented):
Please document any new local variables. Should they be added to FDR as well?

lukecologne commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2003,19 +2005,15 @@ var A320_Neo_UpperECAM;
         getN1GaugeValue() {
             const engineId = (this.index + 1);
-            const value = SimVar.GetSimVarValue("ENG N1 RPM:" + engineId, "percent");
+            let value = SimVar.GetSimVarValue("L:A32NX_ENGINE_N1:" + engineId, "percent");
+            if (value < 0.0) {
+                value = 0;
+            }
             return value;
```
```suggestion
            return Math.max(SimVar.GetSimVarValue("L:A32NX_ENGINE_N1:" + engineId, "percent"), 0);
```

lukecologne commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2003,19 +2005,15 @@ var A320_Neo_UpperECAM;
-                egtImbalance = parseInt(0);
-            }
-            const value = SimVar.GetSimVarValue("L:A32NX_ENGINE_EGT:" + engineId, "celsius") - egtImbalance;
+            const value = SimVar.GetSimVarValue("L:A32NX_ENGINE_EGT:" + engineId, "celsius");
             return value;
```
```suggestion
            return SimVar.GetSimVarValue("L:A32NX_ENGINE_EGT:" + engineId, "celsius");
```

lukecologne commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -1,3 +1,5 @@
+// TazX - 06/08/2021: Engine imbalance has been hardcoded in the Wasm. All references have been eliminated.
```
You don't need to leave comments on what you did, `git blame` can do that

Taz5150 replied:
Deleted comments on both Upper and Lower ECAM js files.

lukecologne commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2198,15 +2196,8 @@ var A320_Neo_UpperECAM;
-            const name = "ENG N2 RPM:" + _engine;
-            let percent = SimVar.GetSimVarValue(name, "percent") - n2Imbalance;
+            const name = "L:A32NX_ENGINE_N2:" + _engine;
+            let percent = SimVar.GetSimVarValue(name, "percent");
```
Just directly return, and use `Math.max()` like I commented above

Taz5150 replied:
Done!

lukecologne commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2198,19 +2189,7 @@ var A320_Neo_UpperECAM;
-                percent = 0;
-            }
-            return percent;
+			return Math.max(SimVar.GetSimVarValue("L:A32NX_ENGINE_N2:" + engineId, "percent"), 0);
```
Looks like my code suggestion had an indentation error, sorry for that

Taz5150 replied:
No problem. Just solved it!! :)

aguther commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2198,19 +2189,7 @@ var A320_Neo_UpperECAM;
-                percent = 0;
-            }
-            return percent;
+            return Math.max(SimVar.GetSimVarValue("L:A32NX_ENGINE_N2:" + engineId, "percent"), 0);
```
```suggestion
            const engineId = this.index + 1;
            return Math.max(SimVar.GetSimVarValue("L:A32NX_ENGINE_N2:" + engineId, "percent"), 0);
```

aguther commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2198,19 +2189,8 @@ var A320_Neo_UpperECAM;
-            }
-            return percent;
+            const engineId = this.index + 1;
+            return Math.max(SimVar.GetSimVarValue("L:A32NX_ENGINE_N2:" + engineId, "percent"), 0);
```
```suggestion
            return Math.max(SimVar.GetSimVarValue("L:A32NX_ENGINE_N2:" + _engine, "percent"), 0);
```

aguther commented on flybywire-aircraft-a320-neo/html_ui/Pages/VCockpit/Instruments/Airliners/FlyByWire_A320_Neo/EICAS/ECAM/A320_Neo_UpperECAM.js:
```diff
@@ -2198,19 +2189,8 @@ var A320_Neo_UpperECAM;
-                percent = 0;
-            }
-            return percent;
+            const engineId = this.index + 1;
```
```suggestion
```

aguther commented on src/fdr2csv/src/main.cpp:
```diff
@@ -14,7 +14,7 @@
 
 // IMPORTANT: this constant needs to increased with every interface change
-const uint64_t INTERFACE_VERSION = 6;
+const uint64_t INTERFACE_VERSION = 7;
```
```suggestion
const uint64_t INTERFACE_VERSION = 8;
```

aguther commented on src/fbw/src/FlightDataRecorder.h:
```diff
@@ -12,7 +12,7 @@
  public:
   // IMPORTANT: this constant needs to increased with every interface change
-  const uint64_t INTERFACE_VERSION = 6;
+  const uint64_t INTERFACE_VERSION = 7;
```
```suggestion
  const uint64_t INTERFACE_VERSION = 8;
```

Samueleonard commented:
Quality Assurance Tester Report

Discord: 1samueleonard1
Object of testing: (#5081)
Tier of Testing: (1)
Date: (12/06/2021)

Testing Process:

. loaded up and set up as normal
. started engines and observed NORMAL CONDITIONS - looks fine
. shut down engines and observed - looks fine however EGT doesnt drop below about 400 for a few minutes, and then drops immediately to OAT

. changed weather to extreme hot (+40)
. started + shutdown - startup ok,  EGT issue

. changed weather to extreme cold (-40)
. started + shutdown - engines started but never got an AVAIL light on either engine, same issue with EGT above

. moved to high altitude airport (la paz)
. started + shutdown normal conditions - NOT TESTED YET
. repeated with extreme high + low temps - NOT TESTED YET

Negatives: as above + oil psi goes past the max indicator line, but continues to rise (normal?)
Testing Results: (Not Passed)

Conclusions: needs work, as written above
Media:

frankkopp commented:
Quality Assurance Tester Report

Discord: Cdr_Maverick#6475
Object of testing: #5081
Tier of Testing: 1
Date: 12.6.2021

Testing Process:
For testing, check that the start-up procedure is as expected in the figures above for different set of environmental conditions:
High and low altitude airports
High and low ambient temperatures
Other than the process itself:
check that EGT initial temperature starts at ambient temperature
check that the transition from start-up to Idle is smooth (since two separate functions are being used)

1. Normal conditions +15° ==> OK
2. Cold -40° ==> OK (but ECAM does never show AVAIL for engines - N1 stays below 18% (see video)
3. Hot +45° ==> OK 
4. High + Cold:  VNKT 4358ft  -40° ==> OK ( but no AVAIL msg)
5. Low + Hot:   LLMZ -358m (-1250ft) ==> OK

Negatives: 
-Startup OK
- Shutdown EGT stays at one value for about 1min then jumps to OAT
- When very cold ECAM never shows AVAIL for engines. 

Testing Results: ??? - is the cold+NON-AVAIL an issue or realistic?

Conclusions: 
Subjective impression is that the startup works as intended and according to the PR. 
The EGT after shutdown issue found by 1samueleonard1 is confirmed although probably not part of this PR.

Media:
Normal https://youtu.be/uG-vlb8HbnA
Cold https://youtu.be/I2NcpzHZjQw
Hot https://youtu.be/2d5UNxNONkU
High Cold https://youtu.be/8LoEYJNK2fM
Low Hot https://youtu.be/zgrJk90X3FQ

Taz5150 commented:
> Quality Assurance Tester Report
> 
> Discord: 1samueleonard1
> Object of testing: (#5081)
> Tier of Testing: (1)
> Date: (12/06/2021)
> 
> Testing Process:
> 
> . loaded up and set up as normal
> . started engines and observed NORMAL CONDITIONS - looks fine
> . shut down engines and observed - looks fine however EGT doesnt drop below about 400 for a few minutes, and then drops immediately to OAT
> 
> . changed weather to extreme hot (+40)
> . started + shutdown - startup ok, EGT issue
> 
> . changed weather to extreme cold (-40)
> . started + shutdown - engines started but never got an AVAIL light on either engine, same issue with EGT above
> 
> . moved to high altitude airport (la paz - 13000ft)
> . started + shutdown normal conditions - startup ok, shutdown EGT issue
> . repeated with extreme high + low temps - NOT TESTED YET
> 
> Negatives: as above + oil psi goes past the max indicator line, but continues to rise (normal?)
> Testing Results: (Not Passed)
> 
> Conclusions: needs work, as written above - cannot verify values are correct, would need real pilot
> Media:

Thank you for the test and feedback!

Only engine start and primary parameters have been modeled so far. This means that:

- Engine shutdown parameters can be (will be) all over the place.
- Secondary engine parameters (everything within the Engine page in the SD) are not modeled and thus maintain original/ vanilla aircraft values.

Regarding AVAIL on the EWD, it is still programmed to appear at N1 = 18%. Although visuals are still not been taken care of, I think it would be beneficial to work on the AVAIL issue and include it in this PR. Will do so and probably have it ready today (06/13).

Thank you again for the effort!

Taz5150 commented:
> Quality Assurance Tester Report
> 
> Discord: Cdr_Maverick#6475
> Object of testing: #5081
> Tier of Testing: 1
> Date: 12.6.2021
> 
> Testing Process:
> For testing, check that the start-up procedure is as expected in the figures above for different set of environmental conditions:
> High and low altitude airports
> High and low ambient temperatures
> Other than the process itself:
> check that EGT initial temperature starts at ambient temperature
> check that the transition from start-up to Idle is smooth (since two separate functions are being used)
> 
> 1. Normal conditions +15° ==> OK
> 2. Cold -40° ==> OK (but ECAM does never show AVAIL for engines - N1 stays below 18% (see video)
> 3. Hot +45° ==> OK
> 4. High + Cold:  VNKT 4358ft  -40° ==> OK ( but no AVAIL msg)
> 5. Low + Hot:   LLMZ -358m (-1250ft) ==> OK
> 
> Negatives:
> -Startup OK
> 
> * Shutdown EGT stays at one value for about 1min then jumps to OAT
> * When very cold ECAM never shows AVAIL for engines.
> 
> Testing Results: ??? - is the cold+NON-AVAIL an issue or realistic?
> 
> Conclusions:
> Subjective impression is that the startup works as intended and according to the PR.
> The EGT after shutdown issue found by 1samueleonard1 is confirmed although probably not part of this PR.
> 
> Media:
> Normal https://youtu.be/uG-vlb8HbnA
> Cold https://youtu.be/I2NcpzHZjQw
> Hot https://youtu.be/2d5UNxNONkU
> High Cold https://youtu.be/8LoEYJNK2fM
> Low Hot https://youtu.be/zgrJk90X3FQ

Thank you for the test!

Same comments apply as before. Shutdown needs to be modeled and AVAIL is considered important enough to have it worked on and included in this PR.

Thank you!

Taz5150 commented:
UPDATE: AVAIL is now correctly modeled.

marcman86 commented:
Quality Assurance Tester Report
Discord:marcman86#4907
Object of testing: #5081
Tier of Testing: 1
Date: 13/06/2021

Testing Process: 
- Spawn C&D at EDDP
- TAT 18°C
- Engine start up > OK
- Engine shut down
- TAT -60°C
- Engine start up > Freeze at N1 = 3.1% 
- Respawn
- TAT -60°C
- Engine start up > Freeze at N1 = 3.1% 
- TAT -42°C
- Engine start resumes
- TAT 54°C
- Engine start up > OK

Negatives: 
- N2 digit background change on start up (0% to about 58%) to from black to dark grey (saw that before, so not part of this PR)
- On TAT of -60°C start up get stuck on about N1 3.1%

Testing Results: Passed, as TAT -60°C is very unrealistic i think
Conclusions: Works
Media: -
Hope this helps :-)

frankkopp commented:
Quality Assurance Tester Report

Discord: Cdr_Maverick#6475
Object of testing: #5081
Tier of Testing: 1
Date: 14.6.2021

Testing Process:

Only re-doing the cold test from above:
Cold -40° ==> OK and AVAIL is shown!

Negatives: none
Testing Results: Passed
Conclusions: LGTM

Cryptic56 commented:
Quality Assurance Tester/Trainee Report

Discord: Cryptic#4731
Object of testing: #5081 
Tier of Testing: 1
Date: 14/06/2021

Testing Process: Spawned CD at UEEE
Temperature -40
started APU and then engines
Observed Startup (Good)
Restart
Temperature +50 
started APU and then engines
Observed startup (Good)
Observed shutdown 


Negatives: On shutdown, the EGT stopped decreasing at about 350 degrees then a minute later dropped to OAT
When changing the temperature the engines EGT drops immediately to the OAT rather than slowly decreasing like the APU EGT
 
Testing Results: Passed

Conclusions: A few issues outlined above but other than that, worked great. Big improvement. Great work. 
(Sorry for the brief report, been a little busy this weekend)
